### PR TITLE
[R4R] Update fees for all cdps

### PR DESCRIPTION
**Pull request by jmahess**
This PR does the following:

Updated all CDP fees instead of only updating for risky CDPs
Removed the `HandleNewDebt` method
Removed calls to 'calculateFees` in keeper methods, since it will already have been done in the BeginBlocker
Removed 'Get/Set PreviousBlockTime` method as it was only used in `HandleNewDebt`
Fixed impacted tests

Closes: #424

### CODE REVIEW

**Comment by jmahess** [comment]
Test output:

```$ make test
ok      github.com/kava-labs/kava/app   0.102s
?       github.com/kava-labs/kava/app/params    [no test files]
?       github.com/kava-labs/kava/cli_test      [no test files]
?       github.com/kava-labs/kava/cmd/kvcli     [no test files]
?       github.com/kava-labs/kava/cmd/kvd       [no test files]
?       github.com/kava-labs/kava/rest_test/setup       [no test files]
ok      github.com/kava-labs/kava/x/auction     0.156s
?       github.com/kava-labs/kava/x/auction/client/cli  [no test files]
?       github.com/kava-labs/kava/x/auction/client/rest [no test files]
ok      github.com/kava-labs/kava/x/auction/keeper      0.591s
ok      github.com/kava-labs/kava/x/auction/simulation  0.171s
ok      github.com/kava-labs/kava/x/auction/types       0.110s
ok      github.com/kava-labs/kava/x/bep3        0.407s
?       github.com/kava-labs/kava/x/bep3/client/cli     [no test files]
?       github.com/kava-labs/kava/x/bep3/client/rest    [no test files]
ok      github.com/kava-labs/kava/x/bep3/keeper 0.689s
ok      github.com/kava-labs/kava/x/bep3/simulation     0.110s
ok      github.com/kava-labs/kava/x/bep3/types  0.130s
--- FAIL: TestModuleTestSuite (3.51s)
    --- FAIL: TestModuleTestSuite/TestSeizeSingleCdpWithFees (1.05s)
        abci_test.go:167: 
                Error Trace:    abci_test.go:167
                Error:          Not equal: 
                                expected: types.Int{i:(*big.Int)(0xc000f8e7c0)}
                                actual  : types.Int{i:(*big.Int)(0xc000f8e620)}
                            
                                Diff:
                Test:           TestModuleTestSuite/TestSeizeSingleCdpWithFees
FAIL
FAIL    github.com/kava-labs/kava/x/cdp 3.833s
?       github.com/kava-labs/kava/x/cdp/client/cli      [no test files]
?       github.com/kava-labs/kava/x/cdp/client/rest     [no test files]
ok      github.com/kava-labs/kava/x/cdp/keeper  26.377s
ok      github.com/kava-labs/kava/x/cdp/simulation      0.095s
ok      github.com/kava-labs/kava/x/cdp/types   0.098s
?       github.com/kava-labs/kava/x/kavadist    [no test files]
ok      github.com/kava-labs/kava/x/kavadist/keeper     0.215s
ok      github.com/kava-labs/kava/x/kavadist/simulation 0.092s
ok      github.com/kava-labs/kava/x/kavadist/types      0.119s
ok      github.com/kava-labs/kava/x/pricefeed   0.142s
?       github.com/kava-labs/kava/x/pricefeed/client/cli        [no test files]
?       github.com/kava-labs/kava/x/pricefeed/client/rest       [no test files]
ok      github.com/kava-labs/kava/x/pricefeed/keeper    0.145s
ok      github.com/kava-labs/kava/x/pricefeed/simulation        0.095s
ok      github.com/kava-labs/kava/x/pricefeed/types     0.107s
ok      github.com/kava-labs/kava/x/validator-vesting   0.126s
?       github.com/kava-labs/kava/x/validator-vesting/client/cli        [no test files]
?       github.com/kava-labs/kava/x/validator-vesting/client/rest       [no test files]
ok      github.com/kava-labs/kava/x/validator-vesting/keeper    0.096s
ok      github.com/kava-labs/kava/x/validator-vesting/simulation        0.063s
ok      github.com/kava-labs/kava/x/validator-vesting/types     0.060s
FAIL
make: *** [test] Error 1
$ ```

**Comment by jmahess** [comment]
@karzak I have done the following thus far:
- changed the update risky fees method to iterate over all cdps instead of only the risky ones
- removed the handleNewDebt method
- removed the abci calls to update risky cdp fees, and handle new debt; and replaced with a call to update fees for all cdps
- removed on test that was calling handleNewDebt
- added a TODO question comment above the `TestModuleTestSuite/TestSeizeSingleCdpWithFees` test which is now FAILING

Note: This code is NOT ready to be merged. It is at the first stage that you mentioned of doing the updates and getting it to build. Not fixing the tests. I updated the code and got it to build successfully. The aforementioned test is still failing and would need to be fixed. And need to decide whether the removed test should be reincluded and/or modified. And also if we want to add new tests.

File: x/cdp/keeper/fees.go [hunk 1]
```diff
@@ -45,11 +40,9 @@ func (k Keeper) UpdateFeesForRiskyCdps(ctx sdk.Context, collateralDenom string,
 	if priceDivLiqRatio.IsZero() {
 		priceDivLiqRatio = sdk.SmallestDec()
 	}
-	// NOTE - we have a fixed cutoff at 110% - this may or may not be changed in the future
-	normalizedRatio := sdk.OneDec().Quo(priceDivLiqRatio).Mul(sdk.MustNewDecFromStr("1.1"))
 
-	// now iterate over all the cdps based on collateral ratio
-	k.IterateCdpsByCollateralRatio(ctx, collateralDenom, normalizedRatio, func(cdp types.CDP) bool {
+	// now iterate over ALL the cdps
+	k.IterateAllCdps(ctx, func(cdp types.CDP) bool {
```
Comment (karzak) [inline x/cdp/keeper/fees.go:45]: This does not handle the case where the accumulated fees are rounded down to zero. Further, it needs to replicate the the functionality that was done previously by `HandleNewDebt`.

For each CDP:
1. Caluclate the additional fees (this is done)
2. Check if zero
3a. If zero, return false without updating cdp state
3b. If not zero, you need to mint debt coins to the cdp account and mint surplus coins divided between the liquidator and savings module accounts. Read the method from handle new debt and replicate it in the iteration loop here.  
3c. Before minting any coins, check if `newFeesSurplus` and `newFeesSavings` is rounded down to zero. If it is, return false without updating state.  
4. Update state at end of loop (currently, this is missing a call to keeper.SetCdp).

Comment (jmahess) [reply x/cdp/keeper/fees.go:34]: 🆗

**Comment by jmahess** [comment]
Builds successfully. Several tests are failing.

File: x/cdp/keeper/fees.go [hunk 2]
```diff
@@ -27,44 +27,52 @@ func (k Keeper) CalculateFees(ctx sdk.Context, principal sdk.Coins, periods sdk.
...
+		// now add the new fees fees to the accumulated fees for the cdp
+		cdp.AccumulatedFees = cdp.AccumulatedFees.Add(newFees...)
 
 		// and set the fees updated time to the current block time since we just updated it
 		cdp.FeesUpdated = ctx.BlockTime()
 		collateralToDebtRatio := k.CalculateCollateralToDebtRatio(ctx, cdp.Collateral, cdp.Principal.Add(cdp.AccumulatedFees...))
 		k.RemoveCdpCollateralRatioIndex(ctx, cdp.Collateral[0].Denom, cdp.ID, oldCollateralToDebtRatio)
 		k.SetCdpAndCollateralRatioIndex(ctx, cdp, collateralToDebtRatio)
+
+		k.SetCDP(ctx, cdp)
```
Comment (karzak) [inline x/cdp/keeper/fees.go:74]: Ah, just realized that `SetCdpAndCollateralRatioIndex` already calls `SetCDP`, so we don't need this line.

File: x/cdp/keeper/fees.go [hunk 3]
```diff
@@ -27,44 +27,52 @@ func (k Keeper) CalculateFees(ctx sdk.Context, principal sdk.Coins, periods sdk.
...
+		feeCoins := sdk.NewCoins(sdk.NewCoin(principalDenom, previousDebt))
+		k.SetTotalPrincipal(ctx, collateralDenom, principalDenom, feeCoins.Add(newFees...).AmountOf(principalDenom))
+
+		// mint surplus coins divided between the liquidator and savings module accounts.
+		k.supplyKeeper.MintCoins(ctx, types.LiquidatorMacc, sdk.NewCoins(sdk.NewCoin(principalDenom, newFeesSurplus)))
+		k.supplyKeeper.MintCoins(ctx, types.SavingsRateMacc, sdk.NewCoins(sdk.NewCoin(principalDenom, newFeesSavings)))
+
+		// TODO QUESTION DO WE STILL NEED TO UPDATE THE ACCUMULATED FEES SINCE WE ALREADY ADDED TO THE PRINCIPAL?
+		// now add the new fees fees to the accumulated fees for the cdp
+		cdp.AccumulatedFees = cdp.AccumulatedFees.Add(newFees...)
```
Comment (karzak) [inline x/cdp/keeper/fees.go:66]: yes, the fees need to be set on the cdp itself, so when a user goes to repay debt, the fees are directly on the object. There is no way to get the 'fees owed' by a cdp by looking at TotalPrincipal or some other field.

File: x/cdp/keeper/fees.go (hunk 2 above)
Comment (jmahess) [reply x/cdp/keeper/fees.go:74]: 🆗  I will remove this

File: x/cdp/keeper/fees.go (hunk 3 above)
Comment (jmahess) [reply x/cdp/keeper/fees.go:66]: ok makes sense. I will remove the todo comment

File: x/cdp/keeper/fees.go (hunk 2 above)
Comment (jmahess) [reply x/cdp/keeper/fees.go]: done

File: x/cdp/keeper/fees.go (hunk 3 above)
Comment (jmahess) [reply x/cdp/keeper/fees.go:65]: done

**Comment by jmahess** [comment]
CDP test output:
```$ go test ./...
ok  	github.com/kava-labs/kava/x/cdp	1.005s
?   	github.com/kava-labs/kava/x/cdp/client/cli	[no test files]
?   	github.com/kava-labs/kava/x/cdp/client/rest	[no test files]
ok  	github.com/kava-labs/kava/x/cdp/keeper	22.987s
ok  	github.com/kava-labs/kava/x/cdp/simulation	0.077s
ok  	github.com/kava-labs/kava/x/cdp/types	(cached)```

**Pull request merged**